src/acme/solvers/tls_alpn.rs

```diff
@@ -0,0 +1,714 @@
+                );
+                openssl_sys::ssl_select_cert_result_t_ssl_select_cert_error
+            }
+        }
+    }
+}
+
+impl SslClientHello {
+    fn connection(&self) -> &ngx_connection_t {
+        unsafe {
+            SSL_get_ex_data(self.ssl(), nginx_sys::ngx_ssl_connection_index)
+                .cast::<ngx_connection_t>()
+                .as_ref()
+                .expect("SSL always has an associated ngx_connection_t")
+        }
+    }
+
+    /// Common part of the ClientHello handler.
+    fn handler(&self) -> Result<bool, Error> {
+        use openssl_sys::{
+            SSL_CTX_get_options, SSL_CTX_get_verify_callback, SSL_CTX_get_verify_mode,
+            SSL_clear_options, SSL_get_options, SSL_set_SSL_CTX, SSL_set_options, SSL_set_verify,
+        };
+
+        if self
+            .extension(TLSEXT_TYPE_application_layer_protocol_negotiation)
+            .map(TlsAlpnIter::new)
+            .transpose()?
+            .map(|mut it| it.any(|x| x == b"acme-tls/1"))
+            != Some(true)
+        {
+            return Ok(false);
+        };
+
+        let amcf = unsafe { get_acme_main_conf(self.connection()) }.ok_or(Error::Configuration)?;
+
+        let ssl_ctx = amcf.ssl.as_ref().ctx.cast::<SSL_CTX>();
+        if ssl_ctx.is_null() {
+            return Err(Error::Configuration);
+        }
+
+        unsafe {
+            let ssl = self.ssl();
+
+            if SSL_set_SSL_CTX(self.ssl(), ssl_ctx).is_null() {
+                return Err(Error::SetContext);
+            }
+
+            SSL_set_verify(
+                ssl,
+                SSL_CTX_get_verify_mode(ssl_ctx),
+                SSL_CTX_get_verify_callback(ssl_ctx),
+            );
+
+            SSL_clear_options(ssl, SSL_get_options(ssl) & !SSL_CTX_get_options(ssl_ctx));
+            SSL_set_options(ssl, SSL_CTX_get_options(ssl_ctx));
+            SSL_set_options(ssl, openssl::ssl::SslOptions::NO_RENEGOTIATION.bits());
+        }
+
+        Ok(true)
+    }
+}
+
+/// Gets module configuration from a connection.
+///
+/// # Safety
+///
+/// This function must not be called after ngx_http_request_t is created for the connection.
+unsafe fn get_acme_main_conf(c: &ngx_connection_t) -> Option<&AcmeMainConfig> {
+    unsafe {
+        let hc = c.data.cast::<nginx_sys::ngx_http_connection_t>().as_ref()?;
+        crate::HttpAcmeModule::main_conf(hc)
+    }
+}
+
+/// Iterates over all the alpn entries in the buffer.
+struct TlsAlpnIter<'a>(&'a [u8]);
+
+impl<'a> TlsAlpnIter<'a> {
+    pub fn new(buf: &'a [u8]) -> Result<TlsAlpnIter<'a>, Error> {
+        let (len, buf) = buf.split_first_chunk().ok_or(Error::Decode)?;
+
+        if buf.len() < u16::from_be_bytes(*len).into() {
+            return Err(Error::Decode);
+        }
+
+        Ok(Self(buf))
+    }
+}
+
+impl<'a> Iterator for TlsAlpnIter<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (len, mut buf) = self.0.split_first_chunk()?;
```

Comment:
Just a thought.
Even though the compiler is able to derive N to be 1, it is hard to see it by reading the code without reading the prototypes for the functions here.
Can we be more explicit here with how many bytes we are trying to split?